### ISSUE TITLE
AI-Powered Unit Inference for Inventory Items

### ISSUE THREAD

**Issue by akisma** [akisma/CostFX#34]
**Epic**: Agent Intelligence Enhancement  
**Priority**: P2 - Enhancement  
**Dependencies**: Issue #20 (provides training data)

## Product Value

Replace pattern-matching unit inference with AI model that learns from historical data and user corrections. This embodies the mission of "making Dave's intuition available programmatically" - the AI learns that "ribeye" is typically "ea" while "romaine" is "lb", just like an experienced restaurant manager would know.

## Current State (Issue #20)

Pattern-matching approach:
```javascript
// Looks for patterns like "16oz", "2lb", "1gal" in item names
_inferUnitFromSquare(squareData) {
  const name = squareData.item_data?.name || '';
  if (/\d+\s*oz/i.test(name)) return 'oz';
  if (/\d+\s*lb/i.test(name)) return 'lb';
  // ... more patterns
  return 'ea'; // default
}
```

**Limitations**:
- Rigid pattern matching
- No learning from corrections
- Can't handle variations or regional differences
- Requires manual pattern updates

### Training Data Sources
1. Historical inventory items (name, category, actual unit used)
2. User corrections (when user changes inferred unit)
3. Cross-restaurant patterns (privacy-preserving aggregation)
4. Industry databases (public food service data)

### Model Architecture
- Fine-tuned LLM (GPT-4o-mini or similar)
- Input: Item name, category, description, pricing
- Output: Unit + confidence score
- Fallback to pattern matching if confidence < 80%

### Learning Loop
```javascript
// User corrects unit
await inventoryItem.update({ unit: 'lb' }); // Was 'ea'

// System learns from correction
await AIUnitInferenceService.recordCorrection({
  itemName: 'Romaine Lettuce',
  category: 'produce.leafy_greens',
  inferredUnit: 'ea',
  correctUnit: 'lb',
  restaurantType: 'italian'
});

// Model improves for similar items across all restaurants
```

## Success Metrics
- Unit inference accuracy > 95% (vs current ~70%)
- Reduction in manual corrections by 80%
- Cross-restaurant learning effectiveness
- User satisfaction with "just works" experience

### Phase 1: Data Collection
- Track all unit assignments/corrections
- Build training dataset (minimum 1000+ samples)
- Privacy-preserving aggregation across restaurants

### Phase 2: Model Training
- Fine-tune LLM on training data
- Validate against holdout set
- A/B test against pattern matching

### Phase 3: Production Deployment
- Replace pattern matching with AI model
- Implement confidence-based fallback
- Monitor accuracy and retrain monthly

### Phase 4: Continuous Learning
- Online learning from user corrections
- Federated learning across restaurants
- Personalization per restaurant type

## Dave's Intuition Made Programmatic

This feature exemplifies our mission: Dave knows instantly that saffron is measured in ounces and romaine in pounds. Our AI will learn this intuition and make it available to all restaurants, continuously improving from collective experience.

**Estimated Effort**: 3-4 weeks (after Issue #20 provides foundation)  
**Value**: High - Core differentiator for "AI that understands restaurant operations"

**Comment by akisma** [comment]
## 📋 Implementation Plan - AI-Powered Unit Inference (Revised)

**Date:** November 2, 2025  
**Status:** On Hold - Alternative approach being pursued first  
**Branch:** feature/first-actual-ai

---

### Key Decisions Made:

✅ **Data Source:** CSV upload (unified format) - NOT POS-specific  
✅ **Training Data:** Create synthetic dataset from scratch (~1,500-2,000 items)  
✅ **Model:** OpenAI GPT-4o-mini with fine-tuning capability  
✅ **Fallback:** Maintain existing pattern-matching `UnitInferrer` as fallback  
✅ **Learning:** User corrections tracked for continuous improvement  

---

### Revised Architecture

```
CSV Upload (Unified Format)
    ↓
POSDataTransformer
    ├─ Feature Flag: USE_AI_UNIT_INFERENCE
    └─ Route to appropriate inferrer
          ↓
    ┌─────┴─────────┐
    │               │
┌───▼────┐   ┌─────▼────────────────┐
│Pattern │   │ AIUnitInferenceServ  │
│Matching│   │ ├─ Redis Cache       │
│(Fallback)   │ ├─ OpenAI GPT-4o-mini│
└─────────┘   │ ├─ Confidence >80%  │
              │ └─ Fallback to pattern│
              └──────┬───────────────┘
                     │
              ┌──────▼───────────────┐
              │UnitCorrectionTracker │
              │├─ Log corrections    │
              │├─ Export training data│
              │└─ Retrain trigger    │
              └──────┬───────────────┘
                     │
              ┌──────▼──────────────┐
              │Database             │
              │├─ unit_inference_...│
              │└─ inventory_items   │
              └─────────────────────┘
```

---

### Implementation Phases

#### **PHASE 1: Synthetic Training Dataset Creation** (2-3 days)

**Goal:** Create comprehensive CSV training dataset covering diverse restaurant inventory items

**Tasks:**
1. Research restaurant inventory patterns (Square, Toast, Olo, Sysco, US Foods)
2. Generate 1,500-2,000 synthetic inventory items with correct units
3. Cover all categories: produce, proteins, dairy, dry goods, beverages, paper/disposables, cleaning, spices
4. Include naming variations: with/without units, brand names, packaging descriptors
5. Augment with real data from `Roof_Top_Costing_2025_Merged.csv`
6. Validate distribution matches real restaurant inventory patterns

**CSV Format:**
```csv
item_name,category,correct_unit,confidence,notes
"Romaine Lettuce 5lb",produce.leafy_greens,lb,1.0,"Pattern includes weight"
"Whole Chicken",proteins.poultry,ea,0.9,"Count item, no pattern"
"Saffron",spices.premium,oz,0.8,"High-value, small quantity"
```

---

#### **PHASE 2: Data Collection Infrastructure** (2-3 days)

**Goal:** Build system to track user corrections for continuous learning

**Deliverables:**
1. Database migration: `unit_inference_corrections` table
2. `UnitCorrectionTracker` service (record, query, export)
3. InventoryItem model hook to auto-detect unit changes
4. API endpoints:
   - `POST /api/v1/ai/unit-corrections` - Manual correction logging
   - `GET /api/v1/ai/unit-corrections` - Correction history
   - `GET /api/v1/ai/training-data/export` - Export for retraining
5. Frontend UI: confidence display, "Was this correct?" feedback, quick correction dropdown

---

#### **PHASE 3: AI Model Integration** (4-5 days)

**Goal:** Integrate OpenAI GPT-4o-mini with fine-tuning

**Deliverables:**
1. `AIUnitInferenceService` class with OpenAI integration
2. Fine-tuning job creation with synthetic dataset
3. Inference logic with 3-tier fallback:
   - Try Redis cache first
   - Call fine-tuned model
   - Fall back to pattern matching if error/low confidence
4. Feature flag: `USE_AI_UNIT_INFERENCE` (gradual rollout)
5. Update `POSDataTransformer` to use AI service when enabled
6. Redis caching layer (30-day TTL)

---

#### **PHASE 4: Production Deployment & Monitoring** (2 days)

**Goal:** Deploy with cost controls and monitoring

**Deliverables:**
1. Performance optimization: batch inference, rate limiting, timeout handling
2. Cost controls: usage tracking, budget alerts
3. Monitoring dashboard: accuracy metrics, API response times, fallback rates
4. Logging: all inferences, fallbacks, debugging info
5. Documentation: TECHNICAL_DOCUMENTATION.md updates, Swagger docs, user guide

---

#### **PHASE 5: Continuous Improvement** (Ongoing)

**Goal:** Learn from corrections and retrain periodically

**Deliverables:**
1. Manual retraining workflow (export → combine → fine-tune → update)
2. A/B testing framework (future)
3. Per-restaurant personalization (future)

---

### Success Metrics

- ✅ Unit inference accuracy >95% (vs current ~70%)
- ✅ Manual corrections reduced by 80%
- ✅ Response time <500ms for inference
- ✅ API cost <$50/month for 10K inferences
- ✅ User satisfaction with "just works" experience

---

### Technical Notes

**Model Choice:**
- GPT-4o-mini fine-tuning: ~$3/1M training tokens, ~$0.30/1M inference tokens
- Alternative: GPT-4o-mini zero-shot with prompt engineering (cheaper but less accurate)

**Data Source Considerations:**
- Works with unified CSV format only (POS-agnostic)
- Real restaurant data: `Roof_Top_Costing_2025_Merged.csv` (113 lines, 25 unique menu items)
- Need to synthesize underlying inventory items (e.g., "Lobster Fried Rice" → rice, lobster, vegetables, soy sauce)

**Integration Points:**
- `backend/src/services/helpers/UnitInferrer.js` - Current pattern matching (keep as fallback)
- `backend/src/services/POSDataTransformer.js` - Integration point for AI service
- `backend/tests/unit/services/unitInferrer.test.js` - Existing test coverage (20+ tests)

---

### Dependencies

- ✅ Issue #20 (Square Inventory Sync) - COMPLETED
- OpenAI API key with fine-tuning access
- Redis for caching layer
- Database migration capability

---

### Next Steps When Resumed

1. Create synthetic training dataset CSV (1,500-2,000 items)
2. Review dataset with principal engineer for approval
3. Proceed with Phase 2 implementation
4. Obtain OpenAI API key for fine-tuning

---

**Status:** Implementation plan documented. Alternative approach being pursued first. Will return to this issue when prioritized.
